api/envoy/extensions/filters/listener/http_inspector/v3/http_inspector.proto

```diff
@@ -18,4 +20,22 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 message HttpInspector {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.listener.http_inspector.v2.HttpInspector";
+
+  // The number of bytes of the request the filter will initially buffer and
+  // inspect to determine the http version, the filter will iteratively double
+  // the buffer up to max_read_buffer_size. If not specified, the initial buffer is 8KiB.
+  google.protobuf.UInt32Value initial_read_buffer_size = 1
+      [(validate.rules).uint32 = {lt: 65537 gt: 15}];
+
+  // The maximum number of bytes the request filter will buffer and inspect
+  // to determine the http version. If not specified, the maximum buffer is
+  // 8KiB
+  google.protobuf.UInt32Value max_read_buffer_size = 2
+    [(validate.rules).uint32 = {lt: 65537 gt: 15}];
+
+  // If set, then HTTP/1.1 will be assumed if the maximum_read_buffer_size is
+  // reached and the HTTP parser has not reached the end of the request line or
+  // found malformed HTTP. If not specified, no HTTP will be assumed.
```

Comment:
> no HTTP will be assumed

Just to clarify, this means "it will stop detection and not indicate any appProtocol", right?

Reply:
If so, this is a breaking change IMO that would be an issue for us. With this, on upgrade a users traffic would suddenly be mis-detected as Not-HTTP and get routed improperly

Reply:
It is a breaking change to go from hanging to bypassing routing logic. It's changing from fail-closed to fail-open effectively

Reply:
Fair point.  I'll change this to continue to fail-closed by default, and keep "not_found" reserved for "definitely not http", not "may or may not be HTTP".

Reply:
Sorry I don't think I quite follow -- what does "not_found" mean/what you said?

It seems like there are a few options when we exceed the buffer:
1. If its HTTP so far, mark it as HTTP/1.1 (I think there was some discussion that maybe this is impossible to do reliably, but being exhaustive here)
2. Mark it as Not HTTP and continue
3. Drop the connection
4. Hang (current behavior, not good)
5. Have no max buffer

FWIW I suspect in Istio we would want (3).. and a very large max buffer size. Dropping the connection feels terrible but i don't see what other options we have really, I think it would eventually exceed the max header size if it continued anyways (?)

Reply:
Currently, we report http_not_found if we can fully parse the request line and find that it is not HTTP.  I was originally planning on changing this to also report http_not_found in the exceeding buffer case, in which we still haven't 100% determined it is HTTP, but it still might be—a sort of indeterminate state.  This is bad for the reasons you mentioned, so I'm going to change it to report read_error and drop the connection.  64kb should be more than enough for a sane request line.